src/test/java/io/confluent/connect/jdbc/sink/integration/PostgresDatatypeIT.java

```diff
@@ -85,10 +91,90 @@ public void after() throws SQLException {
     log.info("Dropped table");
   }
 
+  /**
+   * Verifies that even when the connector encounters exceptions that would cause a connection
+   * with an invalid transaction, the connector sends only the errant record to the error
+   * reporter and establishes a valid transaction for subsequent correct records to be sent to
+   * the actual database.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testPrimaryKeyConstraintsSendsToErrorReporter() throws Exception {
+    props.put(ERRORS_TOLERANCE_CONFIG, ToleranceType.ALL.value());
+    props.put(DLQ_TOPIC_NAME_CONFIG, DLQ_TOPIC_NAME);
+    props.put(DLQ_TOPIC_REPLICATION_FACTOR_CONFIG, "1");
+    props.put(MAX_RETRIES, "2");
+
+    createTableWithPrimaryKey();
+    connect.configureConnector("jdbc-sink-connector", props);
+    waitForConnectorToStart("jdbc-sink-connector", 1);
+
+    final Schema schema = SchemaBuilder.struct().name("com.example.Person")
+        .field("firstname", Schema.STRING_SCHEMA)
+        .field("lastname", Schema.STRING_SCHEMA)
+        .build();
+    final Struct struct = new Struct(schema)
+        .put("firstname", "Christina")
+        .put("lastname", "Brams");
+
+    String kafkaValue = new String(jsonConverter.fromConnectData(tableName, schema, struct));
+    connect.kafka().produce(tableName, null, kafkaValue);
+    // Send the same record for a PK collision
+    connect.kafka().produce(tableName, null, kafkaValue);
+
+    // Now, create and send another normal record
+    Struct struct2 = new Struct(schema)
+        .put("firstname", "Brams")
+        .put("lastname", "Christina");
+
+    kafkaValue = new String(jsonConverter.fromConnectData(tableName, schema, struct2));
+    connect.kafka().produce(tableName, null, kafkaValue);
+
+    ConsumerRecords<byte[], byte[]> records = connect.kafka().consume(1, CONSUME_MAX_DURATION_MS,
```

Comment:
`consume()` calls `poll()` on the consumer, which returns immediately if there are records already fetched. There could be a case where both records ended up in DLQ, but record1 happened to be fetched while record2 was not. In that case, consume(1) would return just record1 despite both being present. I think to prevent this, you would have to do a consume(2) and verify that it times out.

Reply:
Does it make sense to add assertEquals to validate how many records made into database ? I think that would validate only bad records in the batch went to dlq and rest of those good records got persisted into database.

Reply:
@ncliang Yes, that is correct. There could be more, but the consume is only looking for 1. Yeah, that's a better way to test it, actually.

Reply:
I've added a wait for offsets section, so this will guarantee that all the records have been sent and so consuming from the DLQ topic should be sufficient to guarantee the number of records sent there.

Reply:
I think that this should be sufficient. The call to `connect.kafka().consume(...)` ensures that at least the specified number of records is returned, but can and will return more than that if 1) there are more records already present in the topic before the call to `consume`, and 2) the expected number of records is less than 500 (the consumer's default max batch size).